File: trunk/dotnet/picocontainer/src/Test/Tck/AbstractLazyInstantiationTestCase.cs
/*****************************************************************************
 * Copyright (C) PicoContainer Organization. All rights reserved.            *
 * ------------------------------------------------------------------------- *
 * The software in this package is published under the terms of the BSD      *
 * style license a copy of which has been included with this distribution in *
 * the license.txt file.                                                     *
 *                                                                           *
 * Idea by Rachel Davies, Original code by Aslak Hellesoy and Paul Hammant   *
 * C# port by Maarten Grootendorst                                           *
 *****************************************************************************/

using System;
using PicoContainer.Defaults;
using PicoContainer.Tests.TestModel;
using csUnit;

namespace PicoContainer.Tests.Tck
{
  /// <summary>
  /// Summary description for AbstractLazyInstantiationTestCase.
  /// </summary>
  [TestFixture]
  public abstract class AbstractLazyInstantiationTestCase
  {
    protected abstract MutablePicoContainer createPicoContainer();

    public class Kilroy 
    {
      public Kilroy(Havana havana) 
      {
        havana.graffiti("Kilroy was here");
      }
    }

    public class Havana 
    {
      public String paint = "Clean wall";

      public void graffiti(String paint) 
      {
        this.paint = paint;
      }
    }

    [Test]
    public void testLazyInstantiation() 
    {
      MutablePicoContainer pico = createPicoContainer();

      pico.RegisterComponentImplementation(typeof(Kilroy));
      pico.RegisterComponentImplementation(typeof(Havana));

      Assert.True(pico.GetComponentInstance(typeof(Havana)) == pico.GetComponentInstance(typeof(Havana)));
      Assert.NotNull(pico.GetComponentInstance(typeof(Havana)));
      Assert.Equals("Clean wall", ((Havana) pico.GetComponentInstance(typeof(Havana))).paint);
      Assert.NotNull(pico.GetComponentInstance(typeof(Kilroy)));
      Assert.Equals("Kilroy was here", ((Havana) pico.GetComponentInstance(typeof(Havana))).paint);
    }
  }
}

File: java/picocontainer/src/test/org/picocontainer/tck/AbstractLazyInstantiationTestCase.java
package org.picocontainer.tck;

import junit.framework.TestCase;
import org.picocontainer.RegistrationPicoContainer;
import org.picocontainer.PicoRegistrationException;
import org.picocontainer.PicoInitializationException;

/**
 * @author Aslak Helles&oslash;y
 * @version $Revision$
 */
public abstract class AbstractLazyInstantiationTestCase extends TestCase {

    protected abstract RegistrationPicoContainer createRegistrationPicoContainer();

    public static class Kilroy {
        public Kilroy(Havana havana) {
            havana.graffiti("Kilroy was here");
        }
    }

    public static class Havana {
        public String paint = "Clean wall";

        public void graffiti(String paint) {
            this.paint = paint;
        }
    }

    public void testLazyInstantiation() throws PicoInitializationException, PicoRegistrationException {
        RegistrationPicoContainer pico = createRegistrationPicoContainer();

        pico.registerComponentByClass(Kilroy.class);
        pico.registerComponentByClass(Havana.class);

        assertSame(pico.getComponent(Havana.class), pico.getComponent(Havana.class));
        assertNotNull(pico.getComponent(Havana.class));
        assertEquals("Clean wall", ((Havana) pico.getComponent(Havana.class)).paint);
        assertNotNull(pico.getComponent(Kilroy.class));
        assertEquals("Kilroy was here", ((Havana) pico.getComponent(Havana.class)).paint);
    }
}
